File: java/src/IceSSL/AcceptorI.java
// **********************************************************************
//
// Copyright (c) 2003-2008 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceSSL;

import IceInternal.Network;
import IceInternal.TraceLevels;

final class AcceptorI implements IceInternal.Acceptor
{
    public void
    close()
    {
        if(_traceLevels.network >= 1)
        {
            String s = "stopping to accept ssl connections at " + toString();
            _logger.trace(_traceLevels.networkCat, s);
        }

        javax.net.ssl.SSLServerSocket fd;
        synchronized(this)
        {
            fd = _fd;
            _fd = null;
        }
        if(fd != null)
        {
            try
            {
                fd.close();
            }
            catch(java.io.IOException ex)
            {
                // Ignore.
            }
        }
    }

    public void
    listen()
    {
        // Nothing to do.

        if(_traceLevels.network >= 1)
        {
            String s = "accepting ssl connections at " + toString();
            _logger.trace(_traceLevels.networkCat, s);
        }
    }

    public IceInternal.Transceiver
    accept()
    {
        //
        // The plugin may not be fully initialized.
        //
        if(!_instance.initialized())
        {
            Ice.PluginInitializationException ex = new Ice.PluginInitializationException();
            ex.reason = "IceSSL: plugin is not initialized";
            throw ex;
        }
        
        javax.net.ssl.SSLSocket fd = null;
        try
        {
            fd = (javax.net.ssl.SSLSocket)_fd.accept();

            //
            // Check whether this socket is the result of a call to connectToSelf.
            // Despite the fact that connectToSelf immediately closes the socket,
            // the server-side handshake process does not raise an exception.
            // Furthermore, we can't simply proceed with the regular handshake
            // process because we don't want to pass such a socket to the
            // certificate verifier (if any).
            //
            // In order to detect a call to connectToSelf, we compare the remote
            // address of the newly-accepted socket to that in _connectToSelfAddr.
            //
            java.net.SocketAddress remoteAddr = fd.getRemoteSocketAddress();
            synchronized(this)
            {
                if(remoteAddr.equals(_connectToSelfAddr))
                {
                    try
                    {
                        fd.close();
                    }
                    catch(java.io.IOException e)
                    {
                    }
                    return null;
                }
            }

            fd.getSession();
        }
        catch(javax.net.ssl.SSLException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }

            //
            // Unfortunately, the situation where the cipher suite does not match
            // the certificates is not detected until accept is called. If we were
            // to throw a LocalException, the IncomingConnectionFactory would
            // simply log it and call accept again, resulting in an infinite loop.
            // To avoid this problem, we check for the special case and throw
            // an exception that IncomingConnectionFactory doesn't trap.
            //
            if(ex.getMessage().toLowerCase().startsWith("no available certificate corresponds to the ssl cipher " +
                                                        "suites which are enabled"))
            {
                RuntimeException e = new RuntimeException();
                e.initCause(ex);
                throw e;
            }

            Ice.SecurityException e = new Ice.SecurityException();
            e.initCause(ex);
            throw e;
        }
        catch(java.io.IOException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }

            if(IceInternal.Network.connectionLost(ex))
            {
                throw new Ice.ConnectionLostException();
            }

            Ice.SocketException e = new Ice.SocketException();
            e.initCause(ex);
            throw e;
        }
        catch(RuntimeException ex)
        {
            if(fd != null)
            {
                try
                {
                    fd.close();
                }
                catch(java.io.IOException e)
                {
                }
            }
            throw ex;
        }

        if(_traceLevels.network >= 1)
        {
            _logger.trace(_traceLevels.networkCat, "accepting ssl connection\n" +
                          IceInternal.Network.fdToString(fd));
        }

        return new TransceiverI(_instance, fd, _host, _adapterName);
    }

    public void connectToSelf()
    {
        java.nio.channels.SocketChannel fd = IceInternal.Network.createTcpSocket();
        synchronized(this)
        {
            //
            // connectToSelf is called to wake up the thread blocked in
            // accept. We remember the originating address for use in
            // accept. See accept for details.
            //
            IceInternal.Network.doConnect(fd, _addr);
            _connectToSelfAddr = (java.net.InetSocketAddress)fd.socket().getLocalSocketAddress();
        }
        IceInternal.Network.closeSocket(fd);
    }

    public String
    toString()
    {
        return IceInternal.Network.addrToString(_addr);
    }

    int
    effectivePort()
    {
        return _addr.getPort();
    }

    AcceptorI(Instance instance, String adapterName, String host, int port)
    {
        _instance = instance;
        _traceLevels = instance.traceLevels();
        _adapterName = adapterName;
        _host = host;
        _logger = instance.communicator().getLogger();
        _backlog = instance.communicator().getProperties().getPropertyAsIntWithDefault("Ice.TCP.Backlog", 511);

        try
        {
            javax.net.ssl.SSLServerSocketFactory factory = _instance.context().getServerSocketFactory();
            _addr = Network.getAddressForServer(host, port, _instance.protocolSupport());
            if(_traceLevels.network >= 2)
            {
                String s = "attempting to bind to ssl socket " + toString();
                _logger.trace(_traceLevels.networkCat, s);
            }

            _fd = (javax.net.ssl.SSLServerSocket)factory.createServerSocket(port, _backlog, _addr.getAddress());
            // Retrieve the address to find out to which port we're bound.
            _addr = (java.net.InetSocketAddress)_fd.getLocalSocketAddress();

            int verifyPeer =
                _instance.communicator().getProperties().getPropertyAsIntWithDefault("IceSSL.VerifyPeer", 2);
            if(verifyPeer == 0)
            {
                _fd.setWantClientAuth(false);
                _fd.setNeedClientAuth(false);
            }
            else if(verifyPeer == 1)
            {
                _fd.setWantClientAuth(true);
            }
            else
            {
                _fd.setNeedClientAuth(true);
            }
            _fd.setUseClientMode(false);

            String[] cipherSuites =
                _instance.filterCiphers(_fd.getSupportedCipherSuites(), _fd.getEnabledCipherSuites());
            try
            {
                _fd.setEnabledCipherSuites(cipherSuites);
            }
            catch(IllegalArgumentException ex)
            {
                Ice.SecurityException e = new Ice.SecurityException();
                e.reason = "IceSSL: invalid ciphersuite";
                e.initCause(ex);
                throw e;
            }
            if(_instance.securityTraceLevel() > 0)
            {
                StringBuffer s = new StringBuffer();
                s.append("enabling SSL ciphersuites for server socket " + toString() + ":");
                for(int i = 0; i < cipherSuites.length; ++i)
                {
                    s.append("\n  " + cipherSuites[i]);
                }
                _logger.trace(_instance.securityTraceCategory(), s.toString());
            }

            String[] protocols = _instance.protocols();
            if(protocols != null)
            {
                try
                {
                    _fd.setEnabledProtocols(protocols);
                }
                catch(IllegalArgumentException ex)
                {
                    Ice.SecurityException e = new Ice.SecurityException();
                    e.reason = "IceSSL: invalid protocol";
                    e.initCause(ex);
                    throw e;
                }
            }
        }
        catch(java.io.IOException ex)
        {
            try
            {
                if(_fd != null)
                {
                    _fd.close();
                }
            }
            catch(java.io.IOException e)
            {
            }
            _fd = null;
            Ice.SocketException se = new Ice.SocketException();
            se.initCause(ex);
            throw se;
        }
    }

    protected synchronized void
    finalize()
        throws Throwable
    {
        IceUtilInternal.Assert.FinalizerAssert(_fd == null);

        super.finalize();
    }

    private Instance _instance;
    private TraceLevels _traceLevels;
    private String _adapterName;
    private String _host;
    private Ice.Logger _logger;
    private javax.net.ssl.SSLServerSocket _fd;
    private int _backlog;
    private java.net.InetSocketAddress _addr;
    private java.net.InetSocketAddress _connectToSelfAddr;
}

File: cpp/src/IceSSL/AcceptorI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2009 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceSSL/AcceptorI.h>
#include <IceSSL/Instance.h>
#include <IceSSL/TransceiverI.h>
#include <IceSSL/Util.h>

#include <Ice/Communicator.h>
#include <Ice/Exception.h>
#include <Ice/LocalException.h>
#include <Ice/LoggerUtil.h>
#include <Ice/Network.h>
#include <Ice/Properties.h>

using namespace std;
using namespace Ice;
using namespace IceSSL;

SOCKET
IceSSL::AcceptorI::fd()
{
    return _fd;
}

void
IceSSL::AcceptorI::close()
{
    if(_instance->networkTraceLevel() >= 1)
    {
        Trace out(_logger, _instance->networkTraceCategory());
        out << "stopping to accept ssl connections at " << toString();
    }

    SOCKET fd = _fd;
    _fd = INVALID_SOCKET;
    IceInternal::closeSocket(fd);
}

void
IceSSL::AcceptorI::listen()
{
    try
    {
        IceInternal::doListen(_fd, _backlog);
    }
    catch(...)
    {
        _fd = INVALID_SOCKET;
        throw;
    }

    if(_instance->networkTraceLevel() >= 1)
    {
        Trace out(_logger, _instance->networkTraceCategory());
        out << "accepting ssl connections at " << toString();
    }
}

IceInternal::TransceiverPtr
IceSSL::AcceptorI::accept()
{
    //
    // The plug-in may not be initialized.
    //
    if(!_instance->context())
    {
        PluginInitializationException ex(__FILE__, __LINE__);
        ex.reason = "IceSSL: plug-in is not initialized";
        throw ex;
    }

    SOCKET fd = IceInternal::doAccept(_fd);
    IceInternal::setBlock(fd, false);
    IceInternal::setTcpBufSize(fd, _instance->communicator()->getProperties(), _logger);

    BIO* bio = BIO_new_socket(static_cast<int>(fd), BIO_CLOSE);
    if(!bio)
    {
        IceInternal::closeSocketNoThrow(fd);
        SecurityException ex(__FILE__, __LINE__);
        ex.reason = "openssl failure";
        throw ex;
    }

    SSL* ssl = SSL_new(_instance->context());
    if(!ssl)
    {
        BIO_free(bio); // Also closes the socket.
        SecurityException ex(__FILE__, __LINE__);
        ex.reason = "openssl failure";
        throw ex;
    }
    SSL_set_bio(ssl, bio, bio);

    if(_instance->networkTraceLevel() >= 1)
    {
        Trace out(_logger, _instance->networkTraceCategory());
        out << "attempting to accept ssl connection\n" << IceInternal::fdToString(fd);
    }

    //
    // SSL handshaking is performed in TransceiverI::initialize, since
    // accept must not block.
    //
    return new TransceiverI(_instance, ssl, fd, "", true, true, _adapterName);
}

string
IceSSL::AcceptorI::toString() const
{
    return IceInternal::addrToString(_addr);
}

int
IceSSL::AcceptorI::effectivePort() const
{
    if(_addr.ss_family == AF_INET)
    {
        return ntohs(reinterpret_cast<const sockaddr_in*>(&_addr)->sin_port);
    }
    else
    {
        return ntohs(reinterpret_cast<const sockaddr_in6*>(&_addr)->sin6_port);
    }
}

IceSSL::AcceptorI::AcceptorI(const InstancePtr& instance, const string& adapterName, const string& host, int port) :
    _instance(instance),
    _adapterName(adapterName),
    _logger(instance->communicator()->getLogger())
{
#ifdef SOMAXCONN
    _backlog = instance->communicator()->getProperties()->getPropertyAsIntWithDefault("Ice.TCP.Backlog", SOMAXCONN);
#else
    _backlog = instance->communicator()->getProperties()->getPropertyAsIntWithDefault("Ice.TCP.Backlog", 511);
#endif

    try
    {
        IceInternal::getAddressForServer(host, port, _addr, _instance->protocolSupport());
        _fd = IceInternal::createSocket(false, _addr.ss_family);
        IceInternal::setBlock(_fd, false);
        IceInternal::setTcpBufSize(_fd, _instance->communicator()->getProperties(), _logger);
#ifndef _WIN32
        //
        // Enable SO_REUSEADDR on Unix platforms to allow re-using the
        // socket even if it's in the TIME_WAIT state. On Windows,
        // this doesn't appear to be necessary and enabling
        // SO_REUSEADDR would actually not be a good thing since it
        // allows a second process to bind to an address even it's
        // already bound by another process.
        //
        // TODO: using SO_EXCLUSIVEADDRUSE on Windows would probably
        // be better but it's only supported by recent Windows
        // versions (XP SP2, Windows Server 2003).
        //
        IceInternal::setReuseAddress(_fd, true);
#endif
        if(_instance->networkTraceLevel() >= 2)
        {
            Trace out(_logger, _instance->networkTraceCategory());
            out << "attempting to bind to ssl socket " << toString();
        }
        IceInternal::doBind(_fd, _addr);
    }
    catch(...)
    {
        _fd = INVALID_SOCKET;
        throw;
    }
}

IceSSL::AcceptorI::~AcceptorI()
{
    assert(_fd == INVALID_SOCKET);
}
